primitives/state-machine/src/witness_ext.rs

```diff
@@ -0,0 +1,250 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Substrate backend runing on a trie proof, no_std compatible.
+
+use hash_db::Hasher;
+use sp_externalities::{Externalities, ExtensionStore, Error, Extension};
+use sp_core::storage::{ChildInfo, TrackedStorageKey};
+use sp_std::{any::{TypeId, Any}};
+use sp_std::boxed::Box;
+use sp_std::vec::Vec;
+use sp_trie::MemoryDB;
+use crate::trie_backend::TrieBackend;
+use crate::ext::{ExtInner, ExtInnerMut};
+use crate::overlayed_changes::{OverlayedChanges, NoExtrinsics};
+use crate::{StorageValue, StorageKey};
+
+/// The backend runnig on a trie proof.
+pub struct WitnessExt<H: Hasher> {
```

Comment:
The witness ext is only here to allow storage in cumulus, it does not contain implementation (use Ext code internally).
I put this in state machine, but it can also be moved to https://github.com/paritytech/cumulus/blob/bb3bd995e4a8184d95ece9040d0c3ed55aa92456/runtime/src/validate_block/implementation.rs#L195 , I just found it cleaner to use Exteranalities trait between crate (even if the partial implementation of this trait is far from clean).

Reply:
If we use scope_limited from the externalities crate, we don't need this extra struct here and also don't need any function to access the "modified" data, because we have direct access to the overlayed changes.

Reply:
I see, seems doable :+1: .

Reply:
https://github.com/paritytech/cumulus/blob/ec-witness-ext/runtime/src/validate_block/implementation.rs with the scope_limited usage looks indeed better.